http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/nrs/util/BatchedProducer.java

```diff
@@ -58,10 +58,18 @@ public void read(final ChannelHandlerContext ctx) throws Exception {
         ctx.pipeline().channel().eventLoop().parent().execute(new Runnable() {
             @Override
             public void run() {
-                for (int i = 0; i < batchSize && sequence.get() != eofOn; i++) {
-                    ctx.fireChannelRead(sequence.getAndIncrement());
+                long valueToProduce = sequence.get();
+                for (int i = 0; i < batchSize && valueToProduce != eofOn;) {
+                    if (!sequence.compareAndSet(valueToProduce, valueToProduce + 1)) {
+                        // Another thread already produced this value, try again with the next value.
+                        valueToProduce = sequence.get();
+                        continue;
+                    }
+
+                    ctx.fireChannelRead(valueToProduce);
+                    ++i;
```

Comment:
As discussed, let's look into simplifying this either with a single threaded test or using .getAndIncrement() to cache the value instead of get() that then requires a compareAndSet()

Reply:
Re-wrote it to use a single-threaded executor, which is what I *think* the original author was trying to do by scheduling it on the parent event-loop executor. That way there's no concurrency concerns at all.